Added support for Manim type aliases in Sphinx docs + Added new TypeAliases

chopan050 opened the pull request:
## Overview: What does this pull request change?
Added Manim type aliases to Sphinx Autodoc:
- _**[v0.2]** Custom parser which reads ~the `manim/typing.py` file~ **[v2.0] ALL the modules inside the Manim tree**, generates an Abstract Syntax Tree using `ast`, and collects the necessary info from the type aliases._
  - _On one hand it replaces the previous `manim_type_aliases`. On the other, it allows for obtaining information about the type aliases without any unwanted substitutions or sudden loss of information._
  -  _**Thanks to @JasonGrace2282 for working with me on this!**_
- Imported it into `docs.source.conf` and defined the dictionary `autodoc_type_aliases` containing the necessary mappings to support Manim type aliases in Sphinx docs
- _**[v0.1]** Docs for the `manim.typing` module available!_
  - _**[v1.0]** Now the types are referenced from the rest of the docs! If any function or method has a parameter or return value whose type is one of Manim's type aliases, now it will include an hyperlink to the corresponding docs!_
- _**[v0.1]** Custom Sphinx directive for the `manim.typing` module (see [Sphinx Issue #10785](https://github.com/sphinx-doc/sphinx/issues/10785))_-

Added new useful type aliases:
- `QuadraticBezierPath`, `CubicBezierPath` and `BezierPath` are `(3*N, 3)`, `(4*N, 3)` and `(PPC*N, 3)`-shaped arrays of `Point3D` objects respectively, where each one of the `N` groups of 3, 4 or `PPC` points represents a quadratic, cubic or n-th degree Bézier curve respectively. These are essentially the types for `VMobject.points`.
- `QuadraticSpline`, `CubicSpline` and `Spline` are special cases of the above where all the Bézier curves in the array are connected. They represent the "subpaths" used internally by `VMobject`s.
- For the sake of completeness, also added the missing `InternalPoint2D_Array`, `Point2D_Array`, `Vector2` and `CubicBezierPoints_Array` type aliases.
- _**[v1.0]** Added `GrayscaleImage`, `RGBImage` and `RGBAImage`._

_**[v1.0]** Added the missing docstrings to all the type aliases._

_**[v2.1]** Added missing documentation for the modules `manim.utils.docbuild`, `manim.utils.testing` and all their submodules._

_**[v2.1]** Sorted alphabetically the modules in the "Utilities and other modules" entry inside the "Reference manual" in the docs._

_**[v0.1] Huge thanks to MrDiver for adding all the missing `__all__` exports from the Manim files!**_

_**[v0.1]** Updated Sphinx to version 7.2.6._

_**[v0.3]** JasonGrace2282 added the dependency `freeglu3-dev` to prevent Ubuntu tests from crashing with `ImportError: module "GLU" not found`_

## Motivation and Explanation: Why and how do your changes improve the library?
Currently Manim type aliases are not supported by Sphinx, which leads to visual clutter when generating the docs. This PR attempts to fix that.

Before|After
--|--
![bild](https://github.com/ManimCommunity/manim/assets/49853152/92a10d3b-4d14-4086-bbff-931511e4b189) | ![image](https://github.com/ManimCommunity/manim/assets/49853152/3fa63cc5-08ff-4fd3-98f8-1500f44015bf)

![image](https://github.com/ManimCommunity/manim/assets/49853152/2cb84f1a-7da9-4ec4-a18e-06483c23834f)
![image](https://github.com/ManimCommunity/manim/assets/49853152/55192e07-f07b-431c-beea-b911dd67e4d5)

## Links to added or changed documentation pages
- [`manim.typing`](
https://manimce--3484.org.readthedocs.build/en/3484/reference/manim.typing.html)
- [`manim.mobject.mobject`](
https://manimce--3484.org.readthedocs.build/en/3484/reference/manim.mobject.mobject.html)
- [`manim.utils.color.core`](
https://manimce--3484.org.readthedocs.build/en/3484/reference/manim.utils.color.core.html)
- [`manim.utils.docbuild`](
https://manimce--3484.org.readthedocs.build/en/3484/reference/manim.utils.docbuild.html)
- [`manim.utils.testing`](
https://manimce--3484.org.readthedocs.build/en/3484/reference/manim.utils.testing.html)
- [`manim.utils.bezier` as an example](https://manimce--3484.org.readthedocs.build/en/3484/reference/manim.utils.bezier.html)
- basically everything which makes use of type aliases

## Reviewer Checklist
- [x] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [x] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested


-> https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_type_aliases

MrDiver commented:
Current state, the typing module is now added to the documentation but it's contents are not shown because the module members are not written in caps

MrDiver commented on manim/cli/render/global_options.py:
```diff
@@ -6,6 +6,11 @@
 
 from ... import logger
 
+if TYPE_CHECKING:
+    from cloup._option_groups import OptionGroupDecorator
```
```suggestion
```

chopan050 commented:
UPDATES:

- Thanks to MrDiver for including all the missing `__all__` exports in Manim!
- I marked this as a draft, as there's still a lot of progress and changes to do.
- Adding docs for the type aliases is not going to be straightforward. See [Sphinx Issue #10785](https://github.com/sphinx-doc/sphinx/issues/10785)
- A workaround for us would be to implement our own directive for Autodoc. I already started a draft for that.
- You can see how the typing module docs currently look in the link in the main post.

chopan050 commented:
_**[UPDATE v0.2]**_
- **@JasonGrace2282 has done a lot for this PR! Kudos to him!**
- We made a custom parsing function, `get_typing_docs()` in `manim/utils/docbuild/module_parsing.py`, which reads the `manim/typing.py` file, generates an Abstract Syntax Tree using `ast`, and collects the necessary info from the type aliases.
  - On one hand it replaces the previous manim_type_aliases. On the other, it allows for obtaining information about the type aliases without any unwanted substitutions or sudden loss of information.

JasonGrace2282 commented:
![image](https://github.com/ManimCommunity/manim/assets/110117391/e24533df-d13b-470f-814b-7605367638be)

Subject for a future PR:
Get `ParseableManimColor` to render in docs correctly.

chopan050 commented:
_**[UPDATE v1.0]**_

## *Now it's ready for review!*

- Now references to the type aliases will be linked to the respective documentation! 
![image](https://github.com/ManimCommunity/manim/assets/49853152/85ac9d22-b6cc-4307-a76e-442df0dcdab1)
![image](https://github.com/ManimCommunity/manim/assets/49853152/5ba51d4f-c4cc-4e3d-823a-644cdff82a8a)

- The "Module Attributes" section for `manim.typing` is gone, because it's always been intended to be replaced by our custom directive and parsing!
- Small caveat: you can notice that the types are referred to as "classes", which is technically wrong. However, that's the best thing we could do for this. To make the references work, we have to give Sphinx what it wants: a class definition for the types.

There's still something bothering me. One of the tests is failing in Ubuntu because it's failing to import the `GLU` library, which apparently didn't happen before. If someone knows and is able to fix this issue, I would appreciate it very much!

JasonGrace2282 commented:
Note: to prevent the Linux tests from failing with an import error of `GLU` I added the Linux system dependency of `freeglut3-dev`.

chopan050 commented:
_**[UPDATE v2.0]**_

Now we're parsing and generating type alias documentations not only for `manim.typing`, but for **ALL** Manim modules!

This includes `ParsableManimColor`, `Updater`, `TimeBasedUpdater` and `NonTimeBasedUpdater`! Now they also have their respective "Type Alias" entry and hyperlinks.

The performance hit is minimal, as reading all the modules under the Manim tree takes less than 1 second.

JasonGrace2282 commented on manim/_config/utils.py:
```diff
@@ -20,18 +20,24 @@
+
+if TYPE_CHECKING:
+    from enum import EnumMeta
+
```
```suggestion
```
To keep spacing consistent.

chopan050 replied:
I agree, but that change was forced by `isort` in `pre-commit`. I tried to remove them again now, but `isort` inserted them once again. :man_shrugging:

JasonGrace2282 commented on manim/_config/utils.py:
```diff
@@ -20,18 +20,24 @@
+    from enum import EnumMeta
+
+    from typing_extensions import Self
+
```
```suggestion
```

JasonGrace2282 reviewed (approved):
LGTM! Apart from one stylistic thing, this should be good.
However, seeing as I also worked on the PR, it might be better to wait for a third party to review it before merging.

chopan050 commented:
_**[UPDATE v2.1]**_
- The docs for `manim.utils.docbuild` and `manim.utils.testing` were missing, so I added them.
  - Most important was `manim.utils.docbuild`, because the new directive and the file parser are there.
- Sorted alphabetically the entries in the "Utilities and other modules" entry inside the "Reference manual" in the docs (by modifying `docs/source/reference_index/utilities_misc.rst`).

behackl reviewed (approved):
I am very impressed with this PR! These changes make our docs infinitely more useable, kudos!

I'd like to get this merged as soon as possible -- I've just left a bunch of minor comments regarding wording, but other than that this is good to go!

chopan050 commented on manim/typing.py:
```diff
@@ -1,133 +1,462 @@
+
+    .. code-block::
+
+        '''
+        [CATEGORY]
+        <category_name>
+        '''
```
It's not possible to escape the double quotes like that, it raises an error:
```sh
  File "<unknown>", line 89
    \"\"\"
     ^
SyntaxError: unexpected character after line continuation character
```
I would have also preferred using comments such as `# CATEGORY: <category_name>`. However, the `ast` module ignores `#` comments. It only picks up triple-quote multi-line comments, because `ast` considers them as "expressions" (they're strings after all).

behackl replied:
Unfortunate, but as I've said, there is no reason to change the already working solution anyways.

JasonGrace2282 commented on manim/typing.py:
```diff
@@ -1,133 +1,462 @@
+A matrix: ``[[float, ...], [float, ...], ...]``.
+"""
+
+Zeros: TypeAlias = Union[npt.NDArray[PointDType], tuple[tuple[Literal[0], ...], ...]]
```
IIRC it was added in the original typing PR and is not a modification in this PR.

chopan050 replied:
^ this

behackl replied:
Regardless of when it was introduced, I'll probably wonder whether this really is a useful type alias whenever this file is changed next. I've not been working with it, so I can't tell whether it really is useful -- if it is, no action required. If it is not, a follow-up PR might be a good idea. In the scope of this PR there is nothing to do here.

chopan050 replied:
Currently the only two uses of `Zeros` are at `VMobject.get_fill_rgbas()` and `VMobject.get_stroke_rgbas()`. The necessity of this alias is debatable:
```py
    def get_fill_rgbas(self) -> RGBA_Array_Float | Zeros:
        try:
            return self.fill_rgbas
        except AttributeError:
            return np.zeros((1, 4))
```
but, as you mention, if this alias isn't that useful, another PR to handle this would be better :+1:

chopan050 commented on manim/typing.py:
```diff
@@ -1,133 +1,462 @@
+    npt.NDArray[PointDType], tuple[QuadraticBezierPoints, ...]
+]
+"""``shape: (N, 3, 3)``
+An array of `N` `QuadraticBezierPoints` objects: ``[[[float, float, float], [float, float, float], [float, float, float]], ...]``.
```
I did it because I felt weird to also use the math role with PPC for consistency :sweat_smile: but in the end, I did it with both, typing PPC as \text{PPC}.

chopan050 commented on manim/typing.py:
```diff
@@ -1,133 +1,462 @@
-This type alias is mostly made available for internal use and only includes the numpy type.
+VectorND: TypeAlias = Union[npt.NDArray[PointDType], tuple[float, ...]]
+"""``shape (N,)``
+An `N`-D vector: ``[float, ...]``.
```
I agree, I haven't thought about that option :sweat_smile: 
I made the changes!

chopan050 commented on manim/typing.py:
```diff
@@ -1,133 +1,462 @@
+Point3D_Array: TypeAlias = Union[InternalPoint3D_Array, tuple[Point3D, ...]]
+"""``shape: (N, 3)``
+An array of `Point3D` objects: ``[[float, float, float], ...]``.
+(Please refer to the documentation of the function you are using for further type information.)
```
Like with the `Zeros` type alias, these comments were already there when we started this PR.
I don't really know whether this comment is necessary, but because I'd prefer to keep it just in case, I removed the parentheses :+1:

chopan050 commented on manim/typing.py:
```diff
@@ -1,133 +1,462 @@
-RGB_Tuple_Int: TypeAlias = Tuple[int, int, int]
+"""``shape: (3,)``
+A NumPy ndarray of 3 integers between 0 and 255, representing a color in RGB format.
+Its components represent, in order, amounts of: Red light, Green light and Blue light.
```
You're right, I wasn't really satisfied by that either, but didn't know how to word it better.
Your suggestion is much better, I applied it instead! :+1:

chopan050 commented:
_**[UPDATE v2.2]**_
- Handled @behackl's review comments :+1: 
- Added a cautionary note about `Vector2/3/ND` I forgot to add before: not to confuse them with the `Vector`, `Arrow` and `Arrow3D` VMobjects (I also added the same caution in the `Vector` VMobject itself)
- I found out how to add references inside the type alias definition block, so I included that change as well! Especially useful for `ParsableManimColor`, which is an union of many type aliases defined in a different place, `manim.typing`

Viicos commented:
Sorry wasn't able to review, but looks really neat!
